mount.go

```diff
@@ -32,3 +41,189 @@ func isSystemMount(m string) bool {
 
 	return false
 }
+
+func major(dev uint64) int {
+	return int((dev >> 8) & 0xfff)
+}
+
+func minor(dev uint64) int {
+	return int((dev & 0xff) | ((dev >> 12) & 0xfff00))
+}
+
+type device struct {
+	major      int
+	minor      int
+	mountPoint string
+}
+
+var errMountPointNotFound = errors.New("Mount point not found")
+
+// getDeviceForPath gets the underlying device containing the file specified by path.
+// The device type constitutes the major-minor number of the device and the dest mountPoint for the device
+//
+// eg. if /dev/sda1 is mounted on /a/b/c, a call to getDeviceForPath("/a/b/c/file") would return
+//
+//	device {
+//		major : major(/dev/sda1)
+//		manor : minor(/dev/sda1)
+//		mountPoint: /a/b/c
+//	}
+func getDeviceForPath(path string) (device, error) {
+	if path == "" {
+		return device{}, fmt.Errorf("Path cannot be empty")
+	}
+
+	stat := syscall.Stat_t{}
+	err := syscall.Stat(path, &stat)
+	if err != nil {
+		return device{}, err
+	}
+
+	// stat.Dev points to the underlying device containing the file
+	major := major(stat.Dev)
+	minor := minor(stat.Dev)
+
+	mountPoint := path
+
+	if path == "/" {
+		return device{
+			major:      major,
+			minor:      minor,
+			mountPoint: mountPoint,
+		}, nil
+	}
+
+	// We get the mount point by recursively peforming stat on the path
+	// The point where the device changes indicates the mountpoint
+	for {
+		if mountPoint == "/" {
+			return device{}, errMountPointNotFound
+		}
+
+		parentStat := syscall.Stat_t{}
+		parentDir := filepath.Dir(path)
+
+		err := syscall.Lstat(parentDir, &parentStat)
+		if err != nil {
+			return device{}, err
+		}
+
+		if parentStat.Dev != stat.Dev {
+			break
+		}
+
+		mountPoint = parentDir
+		stat = parentStat
+		path = parentDir
+	}
+
+	dev := device{
+		major:      major,
+		minor:      minor,
+		mountPoint: mountPoint,
+	}
+
+	return dev, nil
+}
+
+const (
+	procMountsFile = "/proc/mounts"
+
+	fieldsPerLine = 6
+)
+
+const (
+	procDeviceIndex = iota
+	procPathIndex
+	procTypeIndex
+)
+
+func getDevicePathAndFsType(mountPoint string) (devicePath, fsType string, err error) {
+	if mountPoint == "" {
+		err = fmt.Errorf("Mount point cannot be empty")
+		return
+	}
+
+	var file *os.File
+
+	file, err = os.Open(procMountsFile)
+	if err != nil {
+		return
+	}
+
+	defer file.Close()
+
+	reader := bufio.NewReader(file)
+	for {
+		var line string
+
+		line, err = reader.ReadString('\n')
+		if err == io.EOF {
+			err = fmt.Errorf("Mount %s not found", mountPoint)
+			break
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) != fieldsPerLine {
+			err = fmt.Errorf("Incorrect no of fields (expected %d, got %d)) :%s", fieldsPerLine, len(fields), line)
+			break
+		}
+
+		if mountPoint == fields[procPathIndex] {
+			devicePath = fields[procDeviceIndex]
+			fsType = fields[procTypeIndex]
+			return
+		}
+	}
+
+	return
```

Comment:
If you immediately return an error at the call site, this would presumably need to become a catch-all error.